configure.ac

```diff
@@ -594,20 +594,26 @@ AC_TRY_RUN([
 #ifdef __APPLE__
 #include <mach-o/getsect.h>
 #else
-extern char data_start;
-extern char end;
+#pragma weak __data_start
+#pragma weak _end
+extern int __data_start;
+extern int _end;
 #endif
 
+int data;
+
 int main(void) {
     void *base;
-    unsigned long length;
+    long length;
 #ifdef __APPLE__
     base = (void*) get_etext();
     length = get_end() - get_etext();
 #else
-    base = &data_start;
-    length = (unsigned long) &end  - (unsigned long) &data_start;
+    if (&__data_start == (int*) 0 || &_end == (int*) 0) return 1;
+    base = (void*) &__data_start;
+    length = (long) ((char*) &_end - (char*) &__data_start);
 #endif
+    if (!(base != (void*)0 && length > 0)) return 2;
```

Comment:
The `length>0` seems like a good check.  Do we also want it in `src/init.c`?  Or is it checked further along?

Reply:
A zero-length data segment is technically allowed (although I'm not sure we've ever tested that this works).  The configury test puts something in the data segment, so we expect the size to be nonzero.